### PR TITLE
chore: test environment overrides on all manifest fields shared among workloads 

### PR DESCRIPTION
**Pull request by Lou1415926**
This PR is part of the first step (writing units test) to address #2492 by writing `ApplyEnv` unit tests on all released manifest fields that are shared among workloads.

Some unit tests are marked "FAILED_TEST:.." and commented out. Those tests indicate potential bugs in our current manifest, and should be fixed in a later stage.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 1]
```diff
@@ -0,0 +1,1846 @@
+package manifest
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: copyright header 😂 

Looks like our little script is not running: https://github.com/aws/copilot-cli/blob/mainline/scripts/license.sh w github actions: https://github.com/aws/copilot-cli/pull/2700/checks?check_run_id=3247172643

```
find: unknown predicate `-E'
```
I'll look into sending a fix

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 2]
```diff
@@ -0,0 +1,1846 @@
...
+					BuildString: aws.String("mockBuildTest"),
+				}
+			},
+		},
+		"build not overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: aws.String("mockBuild"),
+				}
+				svc.Environments["test"].ImageConfig.Build = BuildArgsOrString{}
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: nit: can we remove this

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 3]
```diff
@@ -0,0 +1,1846 @@
...
+				svc.ImageConfig.Location = nil
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: nil,
+					BuildArgs: DockerBuildArgs{
+						Dockerfile: aws.String("mockDockerfile"),
+					},
+				}
+			},
+		},
+		"build overridden": {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: how do we know if these test cases should be in `Test_ApplyEnv_Image` vs. `Test_ApplyEnv_Image_Build`?

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 4]
```diff
@@ -0,0 +1,1846 @@
...
+					StartPeriod: nil,
+				}
+			},
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			var inSvc, wantedSvc LoadBalancedWebService
+			inSvc.Environments = map[string]*LoadBalancedWebServiceConfig{
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: I think one risk right now is that our tests is only using one workload type: `LoadBalancedWebServiceConfig` which means the other service types' `ApplyEnv` doesn't end up getting covered.

That's another huge refactor that we have to think about but it'd be really good if we can get to a point where the `ApplyEnv` codepath can be shared across services so we don't have to worry about it.

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 5]
```diff
@@ -0,0 +1,1846 @@
...
+		})
+	}
+}
+
+func Test_ApplyEnv_Image_Build(t *testing.T) {
+	testCases := map[string]struct {
+		inSvc  func(svc *LoadBalancedWebService)
+		wanted func(svc *LoadBalancedWebService)
+	}{
+		"composite fields: build string is overridden if build args is not nil": {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: should this be "exclusive fields" to be consistent w the previous test?

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 6]
```diff
@@ -0,0 +1,1846 @@
...
+			},
+		},
+		"dockerfile not overridden": {
+			inSvc: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildArgs: DockerBuildArgs{
+						Dockerfile: aws.String("mockDockerfile"),
+					},
+				}
+				svc.Environments["test"].ImageConfig.Build = BuildArgsOrString{
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: for the "not overriden" test cases can we remove the `["test"]` lines?

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 7]
```diff
@@ -0,0 +1,1846 @@
...
+			wanted: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: nil,
+					BuildArgs: DockerBuildArgs{
+						Dockerfile: aws.String("mockDockerfile"),
+					},
+				}
+			},
+		},
+		//"FAILED TEST: args overridden": {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: 😱 thank you so much for finding these gaps!

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 8]
```diff
@@ -0,0 +1,1846 @@
...
+					BuildArgs: DockerBuildArgs{
+						CacheFrom: []string{},
+					},
+				}
+			},
+			wanted: func(svc *LoadBalancedWebService) {
+				svc.ImageConfig.Build = BuildArgsOrString{
+					BuildString: nil,
+					BuildArgs: DockerBuildArgs{
+						CacheFrom: []string{},
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: Is this the behavior that we want for `slices`? is it consistent everywhere today?  
I'm just asking because it's not what Compose does :( where they strictly just append to the existing slice.

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 9]
```diff
@@ -0,0 +1,1846 @@
...
+
+			got, err := inSvc.ApplyEnv("test")
+
+			require.NoError(t, err)
+			require.Equal(t, &wantedSvc, got)
+		})
+	}
+}
+
+func Test_ApplyEnv_Entrypoint(t *testing.T) {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: What should be the guideline to our team for when to create a test?
 
We should aim at:
1. Ensuring each field is covered.
2. There should be only one test function where a field is tested. We don't want to make it painful to update the tests.

```
If you have have a multi-type field (such as string or string slice)
   -> Then that should get its own test.
  
Else if you have a struct field (composed of multiple simple types)
   -> Then it should get its own test.
   
Otherwise (a field is just a simple type like "string" or "*string")
   -> The field should be covered in the parent field's test.
```

File: internal/pkg/manifest/workload_applyenv_test.go [hunk 10]
```diff
@@ -0,0 +1,1846 @@
...
+
+import (
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_ApplyEnv_Image(t *testing.T) {
```
Comment (efekarakus) [inline internal/pkg/manifest/workload_applyenv_test.go]: nit: can we rename the tests to follow the format
```suggestion
func TestApplyEnv_Image(t *testing.T) {
```

**Review by efekarakus — COMMENTED** [review]
Awesome thank you so much for the extensive test cases covering! 😍

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: I think it makes sense to update it to
```suggestion
				svc.Environments["test"].ImageConfig= ImageConfig{}
```
This is because sometimes one of the parent field is `nil` and it makes a difference to the test result how far above the struct is initialized :(

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 3 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: Good quesetion!! Glad you asked!!!

In `Test_ApplyEnv_Image`, we should test every member fields of `Image`. If a member field is a built-in type (map, slice, string, etc.), we test 3 cases: 1. override 2. empty value 3. nil. If a member field is a composite type, or a struct, we test two cases 1. override, and 2. nil, and then create another test group `Test_ApplyEnv_Image_Build` for every member field of `Image.Build`

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 4 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: I initially have these tests for each of the workload type but they'd be largely repetitive. This is probably also an indication of the need to refactor.

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 5 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: I meant to differentiate "composite fields" and "exclusive fields". Although in a composite type, the two fields are naturally exclusive, two fields being mutually exclusive doesn't necessarily mean that their parent struct is a composite struct.

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 4 above)
Comment (huanjani) [reply internal/pkg/manifest/workload_applyenv_test.go]: I am in the middle of adding a method to the WorkloadManifest interface, which currently only has ApplyEnv...so I'm doubling the amount of repetition! 😬

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: To better explain this: https://github.com/aws/copilot-cli/blob/ad12a5783597574036da61348af8d515cc2b12a1/internal/pkg/manifest/workload_applyenv_test.go#L1810

When the user specify
```yaml
network:
   vpc:
       placement: some_placement
```
The VPC struct will unmarshaled as 
```go
{
   placement: aws.String("some_placement")
   security_groups: nil
}
```
which results in the failed case in https://github.com/aws/copilot-cli/blob/ad12a5783597574036da61348af8d515cc2b12a1/internal/pkg/manifest/workload_applyenv_test.go#L1810.

However, if the network field is not specified at all, it will be `nil`, and `security_groups` won't be overridden because `Network` is a pointer type and unmarshaled as `nil`.

Comment (efekarakus) [reply internal/pkg/manifest/workload_applyenv_test.go]: I think I understand it! So if the parent field is a struct (like "vpc") we want to make sure that `security_groups` is maintained in the following scenario:
```yaml
network:
  vpc:
    security_groups: ['asdf']

environments:
  test:
     network:
        vpc:
          placement: 'private'
```
that's why you test with empty structs and it makes a lot of sense.


My recommendation is two things:
1. Rename the file to `applyenv_test.go`.
2. Add a big block comment at the top of the file explaining how to write tests here. The guideline should essentially tell me: "OK I've added a new type to the manifest, how should I write test cases for it based on the type"

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 9 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: Here is a "guideline" that I came up with:

When writing tests for a field `F` (e.g. `TestApplyEnv_Image`)

* For each sub field `f`:
    * If `f` has subfields, or if `f` is a composite field: write two test cases for this `f` inside e.g. `TestApplyEnv_Image`:( `f` is non-nil,  `f` is nil)
    * Otherwise, write three test cases for `f`: (nil, non-nil and empty value, non-nil and non-empty value)
* For each sub field `f`:
    * If `f` is mutually exclusive with another field (e.g. it’s possible that `F` is a composite filed):
        * Add test cases that check they are mutually exclusive (one is set nil when the other is not)
* For each sub field `f`:
    * If `f` has subfields, or if `f` is a composite field: write another test group for this field (e.g. `TestApplyEnv_Image_Build`)

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 1 above)
Comment (efekarakus) [reply internal/pkg/manifest/workload_applyenv_test.go]: fixed here: https://github.com/aws/copilot-cli/pull/2706

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 8 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: Good call! All slice override should be "append"

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 6 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: (sort of) discussed in this thread: https://github.com/aws/copilot-cli/pull/2700#discussion_r683619847 (just for the records)

File: internal/pkg/manifest/workload_applyenv_test.go (hunk 8 above)
Comment (Lou1415926) [reply internal/pkg/manifest/workload_applyenv_test.go]: Addressed!

File: internal/pkg/manifest/applyenv_test.go [hunk 11]
```diff
@@ -0,0 +1,1871 @@
...
+
+import (
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/stretchr/testify/require"
+)
+
+/** How to add `ApplyEnv` unit test to a new manifest field:
```
Comment (efekarakus) [inline internal/pkg/manifest/applyenv_test.go:14]: nice

**Review by efekarakus — APPROVED** [review]

**Review by iamhopaul123 — APPROVED** [review]
🚢
